src/gstnlaunch.c

```diff
@@ -335,17 +340,75 @@ message_cb (GstBus * bus, GstMessage * message, gpointer user_data)
   return TRUE;
 }
 
+static GstPadProbeReturn
+sink_events_cb (GstPad * pad, GstPadProbeInfo * probe_info, void *user_data)
+{
+  GstPadProbeReturn ret = GST_PAD_PROBE_OK;
+  GstEvent *event = GST_PAD_PROBE_INFO_EVENT (probe_info);
+  GstScalableBranch *branch = (GstScalableBranch *) user_data;
+
+  switch (GST_EVENT_TYPE (event)) {
+    case GST_EVENT_EOS:
+      GST_DEBUG_OBJECT (branch, "EOS reached the sink");
+      if (branch->player->drain) {
+        GstQuery *query = gst_query_new_drain ();
+        if (!gst_pad_query (pad, query))
+          GST_DEBUG_OBJECT (branch, "drain query failed");
+        gst_query_unref (query);
+      }
+      break;
+    default:
+      break;
+  }
+
+  return ret;
+}
+
+static gboolean
+is_sink_element (GstElement * element)
```

Comment:
Can't you just check if the element has the `GST_ELEMENT_FLAG_SINK` flag?

Reply:
ok need to check its a bin too as pipeline is considered as a sink too :)